helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -299,6 +306,33 @@ private void updateMaintenanceInfo(final HelixDataAccessor accessor) {
     // The following flag is to guarantee that there's only one update per pineline run because we
     // check for whether maintenance recovery could happen twice every pipeline
     _hasMaintenanceSignalChanged = false;
+
+    // If maintenance mode has exited, clear cached timed-out nodes
+    if (!_isMaintenanceModeEnabled) {
+      _timedOutInstanceDuringMaintenance.clear();
+    }
+  }
+
+  private void timeoutNodesDuringMaintenance(final HelixDataAccessor accessor) {
+    // If maintenance mode is enabled and timeout window is specified, filter 'new' live nodes
+    // for timed-out nodes
+    long timeOutWindow = -1;
+    if (_clusterConfig != null) {
+      timeOutWindow = _clusterConfig.getOfflineNodeTimeOutForMaintenanceMode();
+    }
+    if (timeOutWindow >= 0 && isMaintenanceModeEnabled()) {
+      for (String instance : _liveInstanceCache.getPropertyMap().keySet()) {
+        // 1. Check timed-out cache and don't do repeated work;
+        // 2. Check for nodes that didn't exist in the last iteration, because it has been checked;
+        // 3. For all other nodes, check if it's timed-out.
+        // When maintenance mode is first entered, all nodes will be checked as a result.
+        if (!_timedOutInstanceDuringMaintenance.contains(instance)
+            && !_liveInstanceSnapshotForMaintenance.containsKey(instance)
+            && isInstanceTimedOutDuringMaintenance(accessor, instance, timeOutWindow)) {
+          _timedOutInstanceDuringMaintenance.add(instance);
```

Comment:
There could be some issue with this if a long-offline instance (say InstanceA) came back just after the cluster enter maintenance mode and before the next pipeline started.  InstanceA will be in the _liveInstanceSnapshotForMaintenance set (see line 261), so it will skip the timeout check here.

Reply:
For the first time it enters maintenance mode, there will be no snapshot is stored. It's because maintenance mode signal is processed after refreshLiveInstances(). This way, all nodes will be checked when we first enter maintenance mode.

Reply:
_liveInstanceSnapshotForMaintenance will be refreshed in the beginning of every pipeline, it contains all liveInstances (including these that should be timeout-ed), right? Say, if an (long-offline) instance comes back after the first pipeline before the next pipeline, that instance will be included in the _liveInstanceSnapshotForMaintenance and won't be checked here?   I.e, is _liveInstanceSnapshotForMaintenance 
 always equal to getLiveInstances()?  If it is, what is point of keeping a separate cache?

Reply:
No, `_liveInstanceSnapshotForMaintenance` is equal to `getLiveInstances()` from the *last* pipeline. If a long-offline instance coms back after the first pipeline and before the next pipeline, it will *not* be in the snapshot, and will therefore be checked by the timeout logic. Whether this instance fails the check, it will be stored either in the snapshot or in the "timed-out cache", and it will not be checked by the pipeline after.

Reply:
This has been changed. We now rely on the previous calculation result, which makes more sense when reading.